quinn/src/platform/unix.rs

```diff
@@ -373,3 +389,56 @@ pub const BATCH_SIZE: usize = 32;
 
 #[cfg(any(target_os = "macos", target_os = "ios"))]
 pub const BATCH_SIZE: usize = 1;
+
+#[cfg(target_os = "linux")]
+mod gso {
+    use super::*;
+
+    /// Checks whether GSO support is available by setting the UDP_SEGMENT
+    /// option on a socket
+    pub fn supports_gso() -> bool {
+        const GSO_SIZE: libc::c_int = 1500;
+
+        let socket = match std::net::UdpSocket::bind("[::]:0") {
+            Ok(socket) => socket,
+            Err(_) => return false,
+        };
+
+        let rc = unsafe {
+            libc::setsockopt(
+                socket.as_raw_fd(),
+                libc::SOL_UDP,
+                libc::UDP_SEGMENT,
+                &GSO_SIZE as *const _ as _,
+                mem::size_of_val(&GSO_SIZE) as _,
+            )
+        };
+
+        rc != -1
+    }
+
+    pub fn set_segment_size(encoder: &mut cmsg::Encoder, segment_size: u16) {
+        encoder.push(libc::SOL_UDP, libc::UDP_SEGMENT, segment_size);
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+mod gso {
+    use super::*;
+
+    pub fn supports_gso() -> bool {
+        false
+    }
+
+    pub fn set_segment_size(_encoder: &mut cmsg::Encoder, _segment_size: u16) {
+        panic!("Setting a segment size is not supported on current platform");
+    }
+}
```

Comment:
I like this more than I expected. Might want to generalize it in the future, but that's for then.